crypto/tls_prf.c

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2023 Andrew R. Kozlik
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "hmac.h"
+#include "memzero.h"
+#include "sha2.h"
+
+void tls_prf_sha256(const uint8_t *secret, size_t secret_len,
```

Comment:
What do you think about the following alternative implementation? I expect that the memory and time complexity is about the same, but in my opinion, the readability is better.

```
void tls_prf_sha256(const uint8_t *secret, size_t secret_len,
                    const uint8_t *label, size_t label_len, const uint8_t *seed,
                    size_t seed_len, uint8_t *output, size_t out_len) {
  uint8_t a[SHA256_DIGEST_LENGTH] = {0};
  uint8_t result[SHA256_DIGEST_LENGTH] = {0};

  HMAC_SHA256_CTX hmac_ctx = {0}, hmac_ctx_backup = {0};
  hmac_sha256_Init(&hmac_ctx_backup, secret, secret_len);

  // a = HMAC(secret, label + seed)
  hmac_ctx = hmac_ctx_backup;
  hmac_sha256_Update(&hmac_ctx, label, label_len);
  hmac_sha256_Update(&hmac_ctx, seed, seed_len);
  hmac_sha256_Final(&hmac_ctx, a);

  while (1) {
    // result = HMAC(secret, a + label + seed)
    hmac_ctx = hmac_ctx_backup;
    hmac_sha256_Update(&hmac_ctx, a, SHA256_DIGEST_LENGTH);
    hmac_sha256_Update(&hmac_ctx, label, label_len);
    hmac_sha256_Update(&hmac_ctx, seed, seed_len);
    hmac_sha256_Final(&hmac_ctx, result);

    if (out_len <= SHA256_DIGEST_LENGTH) {
      break;
    }

    memcpy(output, result, SHA256_DIGEST_LENGTH);
    output += SHA256_DIGEST_LENGTH;
    out_len -= SHA256_DIGEST_LENGTH;

    // a = HMAC(secret, a)
    hmac_ctx = hmac_ctx_backup;
    hmac_sha256_Update(&hmac_ctx, a, SHA256_DIGEST_LENGTH);
    hmac_sha256_Final(&hmac_ctx, a);
  }

  memcpy(output, result, out_len);
  memzero(&hmac_ctx, sizeof(hmac_ctx));
  memzero(&hmac_ctx_backup, sizeof(hmac_ctx_backup));
  memzero(a, sizeof(a));
  memzero(result, sizeof(result));
}
```

Reply:
In this case, with the current implementation of SHA and HMAC, it looks like it's OK to copy the context `HMAC_SHA256_CTX`. However, in general contexts should be treated as opaque handles whenever possible, because the underlying implementation may change. So I tried to use the existing API to produce an efficient implementation. I think the proposed code costs 2 extra SHA transforms per output block, although that could probably be fixed.

Reply:
> However, in general contexts should be treated as opaque handles whenever possible, because the underlying implementation may change.

This can be solved by implementing `hmac_sha256_Duplicate()`.